File: apps/api/app/schemas/health.py
"""Response contract for the health endpoint."""

from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, Field

ComponentStatus = Literal["up", "down"]
OverallStatus = Literal["ok", "degraded"]


class ComponentHealth(BaseModel):
    """Health of a single dependency the API relies on."""

    status: ComponentStatus
    latency_ms: float | None = Field(
        default=None,
        description="Round-trip time of the probe, in milliseconds.",
    )
    error: str | None = Field(
        default=None,
        description="Failure detail when status is 'down'.",
    )


class HealthReport(BaseModel):
    """Aggregate health of the API and its dependencies."""

    status: OverallStatus
    version: str
    environment: str
    checks: dict[str, ComponentHealth]

File: apps/web/src/lib/health.ts
/**
 * Typed client for the API's health endpoint.
 *
 * Kept out of the component so the fetch/parse logic stays testable and the
 * component only deals with rendering.
 */

import { apiBaseUrl } from "@/lib/config";

export type ComponentStatus = "up" | "down";
export type OverallStatus = "ok" | "degraded";

export interface ComponentHealth {
  status: ComponentStatus;
  latency_ms: number | null;
  error: string | null;
}

export interface HealthReport {
  status: OverallStatus;
  version: string;
  environment: string;
  checks: Record<string, ComponentHealth>;
}

export type HealthResult =
  | { ok: true; report: HealthReport }
  | { ok: false; error: string };

/**
 * Fetch the API health report.
 *
 * A degraded API answers 503 with a valid body, so a non-2xx response is still
 * parsed rather than treated as a transport failure.
 */
export async function fetchHealth(signal?: AbortSignal): Promise<HealthResult> {
  try {
    const response = await fetch(`${apiBaseUrl}/health`, {
      signal,
      cache: "no-store",
    });
    const report = (await response.json()) as HealthReport;
    return { ok: true, report };
  } catch (error) {
    return {
      ok: false,
      error: error instanceof Error ? error.message : "Unknown error",
    };
  }
}
